### DISCUSSION TITLE
Can I load FW image to R5 TCM and boot on TCM?

### DISCUSSION THREAD

**Discussion by gn00618777** [OpenAMP/open-amp#258] [Q&A]
Hi OpenAMP

I have traced open-amp/apps/examples/load_fw.c 
And If I porting this load_fw.c to Linux as Master, and I want to load FW for Cortex-R5 TCM memory and boot on TCM.
One of I need to do is revising line39 of open-amp/lib/include/openamp/remoteproc_loader.h ?
just like below?
`/* Remoteproc loader any address */`
`#define RPROC_LOAD_ANYADDR  ((metal_phys_addr_t)0xFFE00000)`

and I reference the R5 link script open-amp/apps/system/generic/machine/zynqmp_r5/linker_remote.ld

and device-tree 
```
/ {
  reserved-memory {
    #address-cells = <2>;
    #size-cells = <2>;
    ranges;
    rproc_0_dma: rproc@3ed400000 {
    no-map;
    compatible = "shared-dma-pool";
    reg = <0x0 0x3ed40000 0x0 0x100000>;
  };
  rproc_0_reserved: rproc@3ed000000 {
    no-map;
    reg = <0x0 0x3ed00000 0x0 0x40000>;
  };
};

zynqmp-rpu {
  compatible = "xlnx,zynqmp-r5-remoteproc-1.0";
  #address-cells = <2>;
  #size-cells = <2>;
  ranges;
  core_conf = "split";
  r5_0: r5@0 {
    #address-cells = <2>;
    #size-cells = <2>;
    ranges;
    memory-region = <&rproc_0_reserved>, 
    <&rproc_0_dma>;
    pnode-id = <0x7>;
    mboxes = <&ipi_mailbox_rpu0 0>, <&ipi_mailbox_rpu0 1>;
    mbox-names = "tx", "rx";
    tcm_0_a: tcm_0@0 {
      reg = <0x0 0xFFE00000 0x0 0x10000>;
      pnode-id = <0xf>;
    };
    tcm_0_b: tcm_0@1 {
      reg = <0x0 0xFFE20000 0x0 0x10000>;
      pnode-id = <0x10>;
    };
  };
}
```

**Comment by arnopo** [comment]
> Hi OpenAMP
> 
> I have traced open-amp/apps/examples/load_fw.c
> And If I porting this load_fw.c to Linux as Master, and I want to load FW for Cortex-R5 TCM memory and boot on TCM.
> One of I need to do is revising line39 of open-amp/lib/include/openamp/remoteproc_loader.h ?
> just like below?
> `/* Remoteproc loader any address */`
> `#define RPROC_LOAD_ANYADDR ((metal_phys_addr_t)0xFFE00000)`

Could you explain why you need to update this definition? RPROC_LOAD_ANYADDR is used to initialize in OpenAMP structures an undefined  address. So changing this value can lead to unexpected behavior.

I let @edmooring or @bentheredoneththat answer for the parts related to the Zynqmp platform.
